### PR TITLE
Tour block

### PR DESCRIPTION
**Pull request by amieiro**
This PR adds a Gutenberg block with the tours available.

### Frontend:

image

### Backend:

image

## Test

To test this PR, you can add a new "Tour Block" to a page or a post, and you will see a list with the current tours. Click in one of these tours to start showing it. This PR always try to show the tour: it doesn't check if the selected tour is available in the current URL.

If you want to build the block, you need to:
- Install the dependencies, running `npm install`
- Build the block, running `npm run build`

### CODE REVIEW

File: assets/blocks/src/style.scss [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+.wp-block-tour-block-tour-block {
+	background-color: #fff;
+	color: #000;
+	padding: 2px;
+}
+
+ul.tour-block {
+	//list-style: none;
+	//padding: 0 0 0 1rem;
```
Comment (trymebytes) [inline assets/blocks/src/style.scss]: Can we remove this, since it's commented out already?

Comment (amieiro) [reply assets/blocks/src/style.scss]: Done. Thank you!

File: tour.php [hunk 2]
```diff
@@ -616,3 +616,36 @@ function output_tour_button() {
...
+			));
+
+			if ( empty( $posts ) ) {
+				return '<p>' . esc_html__('There are no tours available.', 'tour-block') . '</p>';
+			}
+
+			$list = '<p>' . esc_attr__('These are the available tours:', 'tour-block') . '</p>';
+			$list .= '<ul class="tour-block">';
+			foreach ($posts as $post) {
+				$list .= '<li><span class="pulse" data-tour-id="' . $post->ID . '">' . $post->post_title . '</span></li>';
```
Comment (akirk) [inline tour.php]: Escaping!
```suggestion
				$list .= '<li><span class="pulse" data-tour-id="' . esc_attr( $post->ID ) . '">' . esc_html( $post->post_title ) . '</span></li>';
```

File: tour.php [hunk 3]
```diff
@@ -616,3 +616,36 @@ function output_tour_button() {
...
 } );
+
+/**
+ * Registers the block using the metadata loaded from the `block.json` file.
+ * Behind the scenes, it registers also all assets so they can be enqueued
+ * through the block editor in the corresponding context.
+ *
+ * @see https://developer.wordpress.org/reference/functions/register_block_type/
+ */
+function tour_block_tour_block_block_init() {
```
Comment (akirk) [inline tour.php]: Why double `tour_block`?

File: assets/blocks/src/save.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * editor into `post_content`.
+ *
+ * @see https://developer.wordpress.org/block-editor/reference-guides/block-api/block-edit-save/#save
+ *
+ * @return {Element} Element to render.
+ */
+export default function save() {
+	return (
+		<p { ...useBlockProps.save() }>
+			{ 'Tour Block – hello from the saved content!' }
```
Comment (akirk) [inline assets/blocks/src/save.js]: This seems like a debug default comment. Can we just remove it?

File: tour.php [hunk 5]
```diff
@@ -616,3 +616,36 @@ function output_tour_button() {
...
+			$posts = get_posts(array(
+				'post_type' => 'tour',
+				'order'     => 'ASC',
+			));
+
+			if ( empty( $posts ) ) {
+				return '<p>' . esc_html__('There are no tours available.', 'tour-block') . '</p>';
+			}
+
+			$list = '<p>' . esc_attr__('These are the available tours:', 'tour-block') . '</p>';
```
Comment (akirk) [inline tour.php]: I think we should not put this text here. You can add such a text in a previous block using Gutenberg.

File: tour.php [hunk 6]
```diff
@@ -616,3 +616,36 @@ function output_tour_button() {
...
+	register_block_type( __DIR__ . '/assets/blocks/build', array(
+		'render_callback' => function() {
+
+			$posts = get_posts(array(
+				'post_type' => 'tour',
+				'order'     => 'ASC',
+			));
+
+			if ( empty( $posts ) ) {
+				return '<p>' . esc_html__('There are no tours available.', 'tour-block') . '</p>';
```
Comment (akirk) [inline tour.php]: We should offer a way to change this from Gutenberg. If we need to be able to reduce the list of tours through JS then this text needs to be always rendered, just hidden by default or if we know that there are not tours. Also, the text domain needs to be `tour`

File: tour.php (hunk 2 above)
Comment (akirk) [inline tour.php]: Is the extra span necessary?

File: assets/blocks/src/style.scss [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	padding: 2px;
+}
+
+ul.tour-block {
+	margin: 0.5rem 0;
+	li {
+		padding-left: 1rem;
+		text-decoration: underline;
+	}
+}
```
Comment (akirk) [inline assets/blocks/src/style.scss]: Do we really need this CSS? Wouldn't the theme CSS suffice?

File: assets/blocks/src/edit.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * @see https://developer.wordpress.org/block-editor/reference-guides/block-api/block-edit-save/#edit
+ *
+ * @return {Element} Element to render.
+ */
+export default function Edit() {
+	return (
+		<div { ...useBlockProps() }>
+			<ServerSideRender
+				block="tour-block/tour-block"
```
Comment (akirk) [inline assets/blocks/src/edit.js]: We need a better name:

```suggestion
				block="tour/available-tours"
```

File: assets/blocks/src/block.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+	"version": "0.1.0",
+	"title": "Tour Block",
+	"category": "widgets",
+	"icon": "testimonial",
+	"description": "Show a list of available tours.",
+	"example": {},
+	"supports": {
+		"html": false
+	},
+	"textdomain": "tour-block",
```
Comment (akirk) [inline assets/blocks/src/block.json]:
```suggestion
	"textdomain": "tour",
```

File: assets/blocks/src/block.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
+{
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 3,
+	"name": "tour-block/tour-block",
```
Comment (akirk) [inline assets/blocks/src/block.json]:
```suggestion
	"name": "tour/available-tour",
```

File: assets/blocks/src/block.json [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+	"version": "0.1.0",
+	"title": "Tour Block",
```
Comment (akirk) [inline assets/blocks/src/block.json]:
```suggestion
	"title": "Available Tours",
```

**Review by akirk — CHANGES_REQUESTED** [review]

**Comment by akirk** [comment]
@amieiro I had a couple of thoughts. It would be good if the "empty" text could be customized.

**Comment by akirk** [comment]
A little help from AI:

> This is my php code, I'd like to be able to edit the "no available tours" text from the frontend:
> ```js
> register_block_type( __DIR__ . '/assets/blocks/build', array(
>   'render_callback' => function() {
> 
>     $posts = get_posts(array(
>       'post_type' => 'tour',
>       'order'     => 'ASC',
>     ));
> 
>     if ( empty( $posts ) ) {
>       return '<p>' . esc_html__('There are no tours available.', 'tour-block') . '</p>';
>     }
> 
>     $list = '<p>' . esc_attr__('These are the available tours:', 'tour-block') . '</p>';
>     $list .= '<ul class="tour-block">';
>     foreach ($posts as $post) {
>       $list .= '<li><span class="pulse" data-tour-id="' . esc_attr( $post->ID ) . '">' . esc_html( $post->post_title ) . '</span></li>';
>     }
>     $list .= '</ul>';
> 
>     return $list;
>   }
> ) );
> }
> ```
> .

To make the "no available tours" text editable from the frontend, you can modify your PHP code to use dynamic block attributes. Here's an example of how you can achieve this:

```php
register_block_type( __DIR__ . '/assets/blocks/build', array(
  'render_callback' => function($attributes) {

    $posts = get_posts(array(
      'post_type' => 'tour',
      'order'     => 'ASC',
    ));

    if ( empty( $posts ) ) {
      // Get the "no available tours" text from the block attributes
      $noToursText = isset($attributes['noToursText']) ? $attributes['noToursText'] : esc_html__('There are no tours available.', 'tour-block');
      return '<p>' . $noToursText . '</p>';
    }

    $list = '<p>' . esc_attr__('These are the available tours:', 'tour-block') . '</p>';
    $list .= '<ul class="tour-block">';
    foreach ($posts as $post) {
      $list .= '<li><span class="pulse" data-tour-id="' . esc_attr( $post->ID ) . '">' . esc_html( $post->post_title ) . '</span></li>';
    }
    $list .= '</ul>';

    return $list;
  }
) );
```

In this code, we added a new attribute called `noToursText` to the block. We then check if this attribute is set in the `$attributes` array. If it is set, we use its value as the "no available tours" text; otherwise, we fallback to the default text using the `esc_html__` function.

To make the `noToursText` attribute editable from the frontend, you need to update your block's JavaScript code to include an additional attribute in the `ServerSideRender` component:

```jsx
export default function Edit({ attributes, setAttributes }) {
  return (
    <div { ...useBlockProps() }>
      <InspectorControls>
        <TextControl
          label="No Tours Text"
          value={ attributes.noToursText }
          onChange={( newValue ) => setAttributes({ noToursText: newValue })}
        />
      </InspectorControls>
      <ServerSideRender
        block="tour/available-tours"
        attributes={ attributes }
      />
    </div>
  );
}
```

By adding the `attributes={ attributes }` prop to the `ServerSideRender` component, you pass all the block attributes to the server-side script. This includes the `noToursText` attribute, which will be used to dynamically render the "no available tours" text.

With these changes, you can now edit the "no available tours" text from the frontend by updating the value in the Gutenberg sidebar.

File: tour.php (hunk 3 above)
Comment (amieiro) [reply tour.php]: I have changed it to `tour_available_tours_init`, reflecting the new name

File: assets/blocks/src/save.js (hunk 4 above)
Comment (amieiro) [reply assets/blocks/src/save.js]: Removed.

File: tour.php (hunk 6 above)
Comment (trymebytes) [reply tour.php]: We should use the tour list filter here, to accommodate tours created via JSON.

**Review by trymebytes — CHANGES_REQUESTED** [review]

File: assets/blocks/src/style.scss (hunk 7 above)
Comment (amieiro) [reply assets/blocks/src/style.scss]: Removed.

File: tour.php (hunk 6 above)
Comment (amieiro) [reply tour.php]: Added the filter. I prefer to merge it and then to add this JavaScript code with the shortcode code.

File: tour.php [hunk 12]
```diff
@@ -616,3 +616,35 @@ function output_tour_button() {
...
+ *
+ * @return string
+ */
+function show_tour_list() {
+	if ( empty( apply_filters( 'tour_list', array() ) ) ) {
+		return '<p>' . esc_html__('There are no tours available.', 'tour') . '</p>';
+	}
+	$tour_list = '<ul id="page-tour-list">';
+	foreach ( apply_filters( 'tour_list', array() ) as $tour_id => $tour ) {
+		$tour_list .= '<li><span class="pulse" data-tour-id="' . esc_attr( $tour_id ) . '">' . esc_html( $tour[0]['title'] ) . '</span></li>';
```
Comment (trymebytes) [inline tour.php]: I think we should remove the `pulse` css class here. It is being used in other places for another purpose.

Comment (trymebytes) [inline tour.php:631]: I think we should remove the `pulse` css class here. It is being used in other places for another purpose.

**Review by trymebytes — APPROVED** [review]

**Review by akirk — APPROVED** [review]
Looks good for further iteration.

**Pull request merged**